jobclient/python/__init__.py

```diff
@@ -0,0 +1,206 @@
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+
+import requests
+
+import util
+
+from typing import Dict, Optional
+from urllib.parse import urlencode, urlunparse
+from uuid import UUID
+
+from jobs import Application, Job
+
+_LOG = logging.getLogger(__name__)
+_LOG.addHandler(logging.StreamHandler())
+_LOG.setLevel(logging.DEBUG)
+
+_COOK_IMPERSONATE_HEADER = 'X-Cook-Impersonate'
+
+_DEFAULT_STATUS_UPDATE_INTERVAL_SECONDS = 10
+_DEFAULT_BATCH_REQUEST_SIZE = 32
+_DEFAULT_REQUEST_TIMEOUT_SECONDS = 60
+_DEFAULT_SUBMIT_RETRY_INTERVAL_SECONDS = 10
+_DEFAULT_JOB_ENDPOINT = '/jobs'
+_DEFAULT_DELETE_ENDPOINT = '/rawscheduler'
+
+
+class InstanceDecorator:
+    def decorate(self, inst_args: dict) -> dict:
+        raise NotImplementedError("stub")
+
+
+class JobClient:
+    __netloc: str
+
+    __job_endpoint: str
+    __delete_endpoint: str
+    __group_endpoint: str
+
+    __status_update_interval: int = _DEFAULT_STATUS_UPDATE_INTERVAL_SECONDS
+    __submit_retry_interval: int = _DEFAULT_SUBMIT_RETRY_INTERVAL_SECONDS
+    __batch_request_size: int = _DEFAULT_BATCH_REQUEST_SIZE
+    __request_timeout_seconds: int = _DEFAULT_REQUEST_TIMEOUT_SECONDS
+
+    def __init__(self, host: str, port: int, *,
+                 job_endpoint: str = _DEFAULT_JOB_ENDPOINT,
+                 delete_endpoint: str = _DEFAULT_DELETE_ENDPOINT):
+        """Initialize an instance of the Cook client.
+
+        Parameters
+        ----------
+        :param host: The hostname of the Cook instance.
+        :type host: str
+        :param port: The port at which the Cook instance is listening.
+        :type port: int
+        :param job_endpoint: The endpoint to be reached when scheduling new
+            jobs or getting the status of running jobs. Defaults to `/jobs`.
+        :type job_endpoint: str, optional
+        :param delete_endpoint: The endpoint to be reached when terminating
+            running jobs. Defaults to `/rawscheduler`
+        """
+        self.__netloc = f'{host}:{port}'
+        self.__job_endpoint = job_endpoint
+        self.__delete_endpoint = delete_endpoint
+
+    def submit(self, *,
+               command: str,
+               cpus: float,
+               mem: float,
+               max_retries: int,
+
+               uuid: Optional[UUID] = None,
+               env: Optional[Dict[str, str]] = None,
+               labels: Optional[Dict[str, str]] = None,
+               max_runtime: Optional[int] = None,
+               name: Optional[str] = None,
+               priority: Optional[int] = None,
+               application: Optional[Application] = None) -> UUID:
```

Comment:
Can this just be a string? Is `application.version` required?

Reply:
[This](https://github.com/twosigma/Cook/blob/master/scheduler/src/cook/rest/api.clj#L266-L269) seems to imply that the version is required to be present and non-empty. One alternative would be to allow omission of the version string and fill in with some default value, e.g. `versionless`.

Reply:
That's correct, the API requires a version if you provide an application name. You can, however, not specify either.